File: backend/src/main/scala/domains/submission/model/SubmissionProblemQuery.scala
package domains.submission.model

import io.circe.{Decoder, Encoder}

final case class SubmissionProblemQuery(value: String)

object SubmissionProblemQuery:
  def parse(raw: String): Either[String, SubmissionProblemQuery] =
    val normalized = raw.trim
    if normalized.isEmpty then Left("Submission problem query is required.")
    else Right(SubmissionProblemQuery(normalized))

  given Encoder[SubmissionProblemQuery] = Encoder.encodeString.contramap(_.value)
  given Decoder[SubmissionProblemQuery] = Decoder.decodeString.emap(parse)

File: frontend/src/objects/submission/request/SubmissionProblemQuery.ts
/** 提交列表题目搜索词品牌类型；表示已确认非空的题目过滤输入。 */
export type SubmissionProblemQuery = string & { readonly __brand: 'SubmissionProblemQuery' }

/** 解析 helper 的返回结果；成功携带品牌值，失败携带可展示错误，不抛异常。 */
type ParseResult<T> = { ok: true; value: T } | { ok: false; error: string }

/** 创建提交题目搜索词品牌值；调用前必须确认非空。 */
function createSubmissionProblemQuery(value: string): SubmissionProblemQuery {
  /** 注意：这里的 as 只在 parseSubmissionProblemQuery 校验通过后施加品牌类型。 */
  return value as SubmissionProblemQuery
}

/** 将提交题目搜索词还原为查询字符串；无副作用。 */
export function submissionProblemQueryValue(query: SubmissionProblemQuery): string {
  return query
}

/** 解析提交题目搜索输入；去除首尾空白并拒绝空查询。 */
export function parseSubmissionProblemQuery(rawQuery: string): ParseResult<SubmissionProblemQuery> {
  const normalized = rawQuery.trim()
  if (!normalized) {
    return { ok: false, error: 'Submission problem query is required.' }
  }
  return { ok: true, value: createSubmissionProblemQuery(normalized) }
}
